src/tiled/objecttemplatemodel.h

```diff
@@ -49,21 +49,21 @@ class ObjectTemplateModel : public QAbstractItemModel
 
     QVariant data(const QModelIndex &index, int role = Qt::DisplayRole) const override;
 
-    bool addNewDocument(QString fileName, QString *error);
+    bool addNewDocument(TemplateGroupDocument *document);
     bool saveObjectToDocument(MapObject *object, QString name, int documentIndex);
 
 private:
     ObjectTemplateModel(QObject *parent = nullptr);
-    TemplateDocuments mTemplateDocuments;
+    TemplateDocuments *mTemplateDocuments;
     ObjectTemplate *toObjectTemplate(const QModelIndex &index) const;
     TemplateGroup *toTemplateGroup(const QModelIndex &index) const;
 };
 
-inline void ObjectTemplateModel::setTemplateDocuments(const TemplateDocuments &templateDocuments)
-{ mTemplateDocuments = templateDocuments; }
+inline void ObjectTemplateModel::setTemplateDocuments(TemplateDocuments &templateDocuments)
+{ mTemplateDocuments = &templateDocuments; }
```

Comment:
`mTemplateDocuments` should not be a pointer. Rather, just store the list here by value and when you need it in the `TemplatesDock`, have it call `ObjectTemplateModel::templateDocuments` and remove `TemplatesDock::mTemplateDocuments`. The function should take `const TemplateDocuments &templateDocuments`.

Also, since this affects the model, it should call `beginResetModel` and `endResetModel`, even though currently the function may only be called while the model is not yet used by a view, this could change later on.

It should probably also own the documents, so it should call `qDeleteAll(mTemplateDocuments)` before the assignment, and this should be added to the destructor (instead of doing it in `TemplatesDock`).